File: Camera_macbeth_main/video_processor.py
import cv2
import numpy as np
from macbeth_nonlinear_color_correction import corriger_image
import os
from config import (output_width, output_height, desired_fps, 
                   COLOR_RANGES, COLOR_MASKS, DETECTION_MASK_PATH, 
                   CACHE_FILE_PATH)
from numba import njit
from functools import lru_cache

# Variables globales
mask: np.ndarray | None = None
resized_mask: np.ndarray | None = None
frame_count = 0
last_correction_coefficients = None

def load_mask():
    """
    Charge et prépare le masque pour le traitement vidéo.
    
    Cette fonction tente de charger un masque depuis un fichier. En cas d'échec,
    elle crée un masque blanc par défaut. Le masque est automatiquement redimensionné
    aux dimensions de sortie configurées.
    
    Notes:
        Le masque est stocké dans la variable globale 'mask'
        En cas d'erreur, un masque blanc est créé par défaut
    """
    global mask, resized_mask
    print(f"Tentative de chargement du masque depuis: {DETECTION_MASK_PATH}")

    def create_default_mask() -> None:
        """Crée un masque blanc par défaut aux dimensions de sortie."""
        global mask, resized_mask
        if output_width is None or output_height is None:
            raise ValueError("Les dimensions de sortie doivent être initialisées")
        mask = np.ones((output_height, output_width), dtype=np.uint8) * 255
        resized_mask = mask.copy()
        print("Création d'un masque blanc par défaut")

    try:
        if not os.path.exists(DETECTION_MASK_PATH):
            print(f"ERREUR: Le fichier de masque n'existe pas: {DETECTION_MASK_PATH}")
            create_default_mask()
            return

        # Chargement du masque en niveaux de gris
        mask = cv2.imread(DETECTION_MASK_PATH, 0)
        
        if mask is None:
            print(f"ERREUR: Impossible de charger le masque: {DETECTION_MASK_PATH}")
            create_default_mask()
        else:
            # Pré-calcul du masque redimensionné une seule fois
            original_shape = mask.shape
            resized_mask = cv2.resize(mask, (output_width, output_height))
            print(f"Masque chargé avec succès: {original_shape} et redimensionné à {(output_height, output_width)}")
            
    except Exception as e:
        print(f"ERREUR lors du chargement du masque: {str(e)}")
        create_default_mask()

def setup_video_capture(video_path):
    """
    Configure la capture vidéo avec les paramètres souhaités.
    
    Args:
        video_path (str): Chemin vers le fichier vidéo
    
    Returns:
        cv2.VideoCapture: Objet de capture vidéo configuré
    """
    cap = cv2.VideoCapture(video_path)
    if not cap.isOpened():
        raise ValueError("Erreur : Impossible d'accéder à la vidéo.")
    
    # Configuration du FPS souhaité
    cap.set(cv2.CAP_PROP_FPS, desired_fps)
    return cap

@njit
def apply_mask(frame, mask):
    return cv2.bitwise_and(frame, frame, mask=mask)

@lru_cache(maxsize=1)
def get_resize_matrix(input_shape):
    """Cache la matrice de transformation pour une taille d'entrée donnée"""
    width, height = int(input_shape[1]), int(input_shape[0])
    
    # Vérifier si le redimensionnement est nécessaire
    if (width, height) == (output_width, output_height):
        return None
    
    # Convertir explicitement en float32 pour OpenCV
    src_points = np.array([
        [0, 0],
        [width-1, 0],
        [0, height-1]
    ], dtype=np.float32)
    
    dst_points = np.array([
        [0, 0],
        [output_width-1, 0],
        [0, output_height-1]
    ], dtype=np.float32)
    
    return cv2.getAffineTransform(src_points, dst_points)

@njit
def _apply_single_mask(frame, mask):
    """
    Applique un masque unique à une frame en utilisant Numba pour l'optimisation.
    
    Args:
        frame (np.ndarray): Image à masquer
        mask (np.ndarray): Masque binaire à appliquer
    
    Returns:
        np.ndarray: Image masquée
    """
    return frame * (mask[:, :, None] / 255.0)

def apply_masks_batch(frames, masks):
    """
    Applique un lot de masques à un lot de frames.
    
    Args:
        frames (np.ndarray): Lot d'images à masquer
        masks (np.ndarray): Lot de masques binaires à appliquer
    
    Returns:
        np.ndarray: Lot d'images masquées
    """
    return np.array([_apply_single_mask(frame, mask) 
                    for frame, mask in zip(frames, masks)])

def process_frame(frame_raw, detect_squares):
    """
    Traite une frame individuelle de la vidéo.
    
    Le traitement inclut :
    1. Redimensionnement aux dimensions configurées
    2. Application du masque si disponible
    3. Correction des couleurs via l'algorithme Macbeth
    
    Args:
        frame_raw (np.array): Image brute à traiter (format BGR)
        detect_squares (bool): Si True, détecte les carrés Macbeth, sinon utilise le cache
    
    Returns:
        np.array: Image traitée avec les couleurs corrigées et le masque appliqué
    """
    try:
        # Éviter le redimensionnement si les dimensions sont déjà correctes
        matrix = get_resize_matrix(frame_raw.shape)
        if matrix is None:
            frame_resized = frame_raw
        else:
            frame_resized = cv2.warpAffine(frame_raw, matrix, (output_width, output_height))
        
        # Application vectorisée du masque
        if resized_mask is not None:
            frame_masked = apply_masks_batch(
                np.array([frame_resized]), 
                np.array([resized_mask])
            )[0]
        else:
            frame_masked = frame_resized
        
        # Correction des couleurs avec gestion des erreurs
        frame_corrected = corriger_image(frame_masked, CACHE_FILE_PATH, detect_squares)
        if frame_corrected is None:
            print("Erreur: La correction des couleurs a échoué")
            return frame_masked
            
        return frame_corrected
        
    except Exception as e:
        print(f"Erreur lors du traitement de la frame: {str(e)}")
        return frame_raw

def initialize_color_masks():
    """
    Initialise les masques de couleur avec vérification vectorisée.
    """
    try:
        # Conversion des plages en tableaux NumPy pour un traitement vectorisé
        hsv_ranges = np.array([(hsv_min, hsv_max) for color_name, (hsv_min, hsv_max) in COLOR_RANGES.items()])
        
        # Vérification vectorisée des valeurs HSV
        if not np.all((hsv_ranges[:, 0, 0] >= 0) & (hsv_ranges[:, 1, 0] <= 180)):
            raise ValueError("Valeurs H invalides détectées")
        if not np.all((hsv_ranges[:, :, 1:] >= 0) & (hsv_ranges[:, :, 1:] <= 255)):
            raise ValueError("Valeurs S/V invalides détectées")
        
        # Création vectorisée des masques
        for color_name, (hsv_min, hsv_max) in COLOR_RANGES.items():
            COLOR_MASKS[color_name] = {
                'min': np.array(hsv_min, dtype=np.uint8),
                'max': np.array(hsv_max, dtype=np.uint8)
            }
        print(f"Masques de couleurs initialisés pour {len(COLOR_MASKS)} couleurs")
        
    except Exception as e:
        print(f"Erreur lors de l'initialisation des masques de couleurs: {str(e)}")
        raise

def get_color_mask(color_name):
    """
    Récupère les seuils min et max pour une couleur donnée.
    """
    if color_name not in COLOR_MASKS:
        raise ValueError(f"Couleur non reconnue : {color_name}")
    return COLOR_MASKS[color_name]['min'], COLOR_MASKS[color_name]['max']

@njit
def process_color_ranges():
    """Pré-calcule les plages de couleurs"""
    return np.array([
        (hsv_min, hsv_max) 
        for _, (hsv_min, hsv_max) in COLOR_RANGES.items()
    ])

File: Camera_main_c/VideoProcessor.cs
using System;
using System.Drawing;
using Emgu.CV;
using Emgu.CV.Structure;
using Emgu.CV.CvEnum;
using System.Collections.Generic;
using System.Runtime.Caching;
using System.IO;

public class VideoProcessor
{
    private static Mat mask;
    private static Mat resizedMask;
    private static int frameCount;
    private static double[] lastCorrectionCoefficients;
    private static MemoryCache resizeMatrixCache = MemoryCache.Default;
    private Mat _frame;
    private Mat _hsv;
    private Mat _mask;
    private readonly int[] _hsvMin = new int[] { 0, 0, 0 };
    private readonly int[] _hsvMax = new int[] { 180, 255, 255 };

    public VideoProcessor()
    {
        _frame = new Mat();
        _hsv = new Mat();
        _mask = new Mat();
    }

    public static void LoadMask(string maskPath)
    {
        Console.WriteLine($"Tentative de chargement du masque depuis: {maskPath}");

        try
        {
            if (!File.Exists(maskPath))
            {
                Console.WriteLine($"ERREUR: Le fichier de masque n'existe pas: {maskPath}");
                CreateDefaultMask();
                return;
            }

            mask = CvInvoke.Imread(maskPath, ImreadModes.Grayscale);
            
            if (mask.IsEmpty)
            {
                Console.WriteLine($"ERREUR: Impossible de charger le masque: {maskPath}");
                CreateDefaultMask();
            }
            else
            {
                var originalShape = mask.Size;
                resizedMask = new Mat();
                CvInvoke.Resize(mask, resizedMask, new Size(Config.DISPLAY_WIDTH, Config.DISPLAY_HEIGHT));
                Console.WriteLine($"Masque chargé avec succès: {originalShape} et redimensionné à {(Config.DISPLAY_HEIGHT, Config.DISPLAY_WIDTH)}");
            }
        }
        catch (Exception e)
        {
            Console.WriteLine($"ERREUR lors du chargement du masque: {e.Message}");
            CreateDefaultMask();
        }
    }

    private static void CreateDefaultMask()
    {
        if (Config.DISPLAY_WIDTH <= 0 || Config.DISPLAY_HEIGHT <= 0)
        {
            throw new InvalidOperationException("Les dimensions de sortie doivent être initialisées");
        }

        mask = new Mat(Config.DISPLAY_HEIGHT, Config.DISPLAY_WIDTH, DepthType.Cv8U, 1);
        mask.SetTo(new MCvScalar(255));
        resizedMask = mask.Clone();
        Console.WriteLine("Création d'un masque blanc par défaut");
    }

    public static VideoCapture SetupVideoCapture(string videoPath)
    {
        var capture = new VideoCapture(videoPath);
        if (!capture.IsOpened)
        {
            throw new InvalidOperationException("Erreur : Impossible d'accéder à la vidéo.");
        }

        capture.Set(CapProp.Fps, Config.DESIRED_FPS);
        return capture;
    }

    private static Mat ApplyMask(Mat frame, Mat mask)
    {
        var result = new Mat();
        CvInvoke.BitwiseAnd(frame, frame, result, mask);
        return result;
    }

    private static Mat GetResizeMatrix(Size inputShape)
    {
        string cacheKey = $"resize_matrix_{inputShape.Width}_{inputShape.Height}";
        
        if (resizeMatrixCache.Contains(cacheKey))
        {
            return (Mat)resizeMatrixCache.Get(cacheKey);
        }

        if (inputShape.Width == Config.DISPLAY_WIDTH && inputShape.Height == Config.DISPLAY_HEIGHT)
        {
            return null;
        }

        var srcPoints = new PointF[]
        {
            new PointF(0, 0),
            new PointF(inputShape.Width - 1, 0),
            new PointF(0, inputShape.Height - 1)
        };

        var dstPoints = new PointF[]
        {
            new PointF(0, 0),
            new PointF(Config.DISPLAY_WIDTH - 1, 0),
            new PointF(0, Config.DISPLAY_HEIGHT - 1)
        };

        var matrix = CvInvoke.GetAffineTransform(srcPoints, dstPoints);
        
        var cacheItem = new CacheItem(cacheKey, matrix);
        var policy = new CacheItemPolicy { SlidingExpiration = TimeSpan.FromMinutes(10) };
        resizeMatrixCache.Add(cacheItem, policy);

        return matrix;
    }

    public static Mat ProcessFrame(Mat frameRaw, string cacheFile, bool detectSquares)
    {
        try
        {
            Mat frameResized;
            var matrix = GetResizeMatrix(frameRaw.Size);
            
            if (matrix == null)
            {
                frameResized = frameRaw.Clone();
            }
            else
            {
                frameResized = new Mat();
                CvInvoke.WarpAffine(frameRaw, frameResized, matrix, 
                    new Size(Config.DISPLAY_WIDTH, Config.DISPLAY_HEIGHT));
            }

            Mat frameMasked;
            if (resizedMask != null)
            {
                frameMasked = ApplyMask(frameResized, resizedMask);
            }
            else
            {
                frameMasked = frameResized;
            }

            var frameCorrected = MacbethColorCorrection.CorrigerImage(frameMasked, cacheFile, detectSquares);
            if (frameCorrected == null)
            {
                Console.WriteLine("Erreur: La correction des couleurs a échoué");
                return frameMasked;
            }

            return frameCorrected;
        }
        catch (Exception e)
        {
            Console.WriteLine($"Erreur lors du traitement de la frame: {e.Message}");
            return frameRaw;
        }
    }

    public static void InitializeColorMasks()
    {
        try
        {
            foreach (var colorRange in Config.COLOR_RANGES)
            {
                var (hsvMin, hsvMax) = colorRange.Value;
                
                if (hsvMin[0] < 0 || hsvMax[0] > 180 ||
                    hsvMin[1] < 0 || hsvMax[1] > 255 ||
                    hsvMin[2] < 0 || hsvMax[2] > 255)
                {
                    throw new ArgumentException($"Valeurs HSV invalides pour {colorRange.Key}");
                }

                Config.COLOR_MASKS[colorRange.Key] = new ColorMask
                {
                    Min = new ScalarArray(hsvMin),
                    Max = new ScalarArray(hsvMax)
                };
            }

            Console.WriteLine($"Masques de couleurs initialisés pour {Config.COLOR_MASKS.Count} couleurs");
        }
        catch (Exception e)
        {
            Console.WriteLine($"Erreur lors de l'initialisation des masques de couleurs: {e.Message}");
            throw;
        }
    }

    public static (ScalarArray Min, ScalarArray Max) GetColorMask(string colorName)
    {
        if (!Config.COLOR_MASKS.ContainsKey(colorName))
        {
            throw new ArgumentException($"Couleur non reconnue : {colorName}");
        }
        return (Config.COLOR_MASKS[colorName].Min, Config.COLOR_MASKS[colorName].Max);
    }

    public void ProcessFrame(Mat frame)
    {
        frame.CopyTo(_frame);
        CvInvoke.CvtColor(_frame, _hsv, ColorConversion.Bgr2Hsv);
        
        var correctedMin = new MCvScalar(_hsvMin[0], _hsvMin[1], _hsvMin[2], 0);
        var correctedMax = new MCvScalar(_hsvMax[0], _hsvMax[1], _hsvMax[2], 0);

        CvInvoke.InRange(_hsv, correctedMin, correctedMax, _mask);
    }

    public void Dispose()
    {
        _frame?.Dispose();
        _hsv?.Dispose();
        _mask?.Dispose();
    }
} 